Training tutorial (#2473)
huggingface/diffusers: docs/source/en/tutorials/basic_training.mdx

<!--Copyright 2023 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

[[open-in-colab]]

# Train a diffusion model

> A popular application of diffusion models is unconditional image generation. The model generates an image that resembles the dataset it was trained on. Typically, the best results are obtained from finetuning a pretrained model on a specific dataset. You can find many of these checkpoints on the [Hub](https://huggingface.co/models?library=diffusers&sort=downloads), but if you can't find one you like, you can always train your own!

Reviewer: Most of them would be text-to-image models (especially when sorted by downloads). Maybe we should skip this mention altogether (or stress that unconditional image generation is a "step before" text-to-image models like Stable Diffusion).

Author: Thanks to the new full-text search, we can find [checkpoints](https://huggingface.co/search/full-text?q=unconditional-image-generation&type=model) that include `tag: unconditional-image-generation` in the model card!

> This tutorial will show you how to finetune a [`UNet2DModel`] on a subset of the [Smithsonian Butterflies](https://huggingface.co/datasets/huggan/smithsonian_butterflies_subset) dataset to generate your own 🦋 butterflies 🦋.

Reviewer: ```suggestion
This tutorial will show you how to train a [`UNet2DModel`] from scratch on a subset of the [Smithsonian Butterflies](https://huggingface.co/datasets/huggan/smithsonian_butterflies_subset) dataset to generate your own 🦋 butterflies 🦋.
```

<Tip>

💡 This training tutorial is based on the [Training with 🧨 Diffusers](https://colab.research.google.com/github/huggingface/notebooks/blob/main/diffusers/training_example.ipynb) notebook. For additional details and context about diffusion models like how they work, check out the notebook!

</Tip>

> Before you begin, make sure you have 🤗 Datasets installed to load and preprocess image datasets and 🤗 Accelerate to simplify training on any number of GPUs. The following command will also install [TensorBoard](https://www.tensorflow.org/tensorboard) to visualize training metrics.

Reviewer: Maybe W&B is easier to start with?

Author: I've never really used W&B, so I'm open to either! I'll add a little note that you can also use W&B.

```bash

> !pip install diffusers[training]==0.11.1

Author: Does this need to be set to version `0.11.1`?

Reviewer 2: @patrickvonplaten
I have this question too  - I assume that we will have to set it to a version so that the code always works, but we probably should test the code out with new versions and update from time to time? 

How is this done in transformers notebooks?

Reviewer 3: We usually use the latest stable version of `transformers` there. Probably we wanted to be ultra-concrete to avoid unusual side-effects of version changes but I don't see any potential issues if we rather used the latest stable version of diffusers here.

```

We encourage you to share your model with the community, and in order to do that, you'll need to login to your Hugging Face account (create one [here](https://hf.co/join) if you don't already have one!) and enter your token when prompted:

```py
>>> from huggingface_hub import notebook_login

>>> notebook_login()

> ```

Reviewer: Or use `huggingface-cli login` in a terminal, if they are not using a notebook.

>>> noisy_image = noise_scheduler.add_noise(sample_image, noise, timesteps)

>>> Image.fromarray(((noisy_image.permute(0, 2, 3, 1) + 1.0) * 127.5).type(torch.uint8).numpy()[0])
```

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/noisy_butterfly.png"/>
</div>

The training objective of the model is to predict the noise added to the image. The loss at this step can be calculated by:

```py
>>> import torch.nn.functional as F

>>> noise_pred = model(noisy_image, timesteps).sample
>>> loss = F.mse_loss(noise_pred, noise)
```

## Train the model

By now, you have most of the pieces to start training the model and all that's missing is putting everything together.

First, you'll need an optimizer and a learning rate scheduler:

```py
>>> from diffusers.optimization import get_cosine_schedule_with_warmup

>>> optimizer = torch.optim.AdamW(model.parameters(), lr=config.learning_rate)
>>> lr_scheduler = get_cosine_schedule_with_warmup(
...     optimizer=optimizer,
...     num_warmup_steps=config.lr_warmup_steps,
...     num_training_steps=(len(train_dataloader) * config.num_epochs),
... )
```

> Then, you'll need to way to evaluate the model. For evaluation, you can use the [`DDPMPipeline`] to generate a batch of sample images and save it as a grid:

Reviewer: ```suggestion
Then, you'll need a way to evaluate the model. For evaluation, you can use the [`DDPMPipeline`] to generate a batch of sample images and save it as a grid:
```

>>> def make_grid(images, rows, cols):
...     w, h = images[0].size
...     grid = Image.new("RGB", size=(cols * w, rows * h))
...     for i, image in enumerate(images):
...         grid.paste(image, box=(i % cols * w, i // cols * h))
...     return grid

>>> def evaluate(config, epoch, pipeline):
...     # Sample some images from random noise (this is the backward diffusion process).
...     # The default pipeline output type is `List[PIL.Image]`
...     images = pipeline(
...         batch_size=config.eval_batch_size,
...         generator=torch.manual_seed(config.seed),
...     ).images

...     # Make a grid out of the images
...     image_grid = make_grid(images, rows=4, cols=4)

...     # Save the images
...     test_dir = os.path.join(config.output_dir, "samples")
...     os.makedirs(test_dir, exist_ok=True)
...     image_grid.save(f"{test_dir}/{epoch:04d}.png")
```

Wrap all these components together and write a training function with 🤗 Accelerate for easy TensorBoard logging, gradient accumulation, and mixed precision training. To upload the model to the Hub, you can also write a function to get your repository name and information and then push it to the Hub.

<Tip>

> 💡 The training loop below may look intimidating and long, but it'll be worth it later when you launch your training in just one line of code! If you can't wait and want to start generating images, feel free to copy and run the code below. You can always come back and examine the training loop more in-depth later, like when you're waiting for your model to finish training. 🤗

Reviewer: ❤️

This tutorial will show you how to train a [`UNet2DModel`] from scratch on a subset of the [Smithsonian Butterflies](https://huggingface.co/datasets/huggan/smithsonian_butterflies_subset) dataset to generate your own 🦋 butterflies 🦋.

<Tip>

💡 This training tutorial is based on the [Training with 🧨 Diffusers](https://colab.research.google.com/github/huggingface/notebooks/blob/main/diffusers/training_example.ipynb) notebook. For additional details and context about diffusion models like how they work, check out the notebook!

</Tip>

Before you begin, make sure you have 🤗 Datasets installed to load and preprocess image datasets and 🤗 Accelerate to simplify training on any number of GPUs. The following command will also install [TensorBoard](https://www.tensorflow.org/tensorboard) to visualize training metrics (you can also use [Weights & Biases](https://docs.wandb.ai/) to track your training).

```bash

> !pip install diffusers[training]==0.13.0

Reviewer 4: ```suggestion
!pip install diffusers[training]
```

@stevhliu should we maybe not pin it so that it stays up to date? :-)
